### ISSUE TITLE
better lfo's, with more functionality:  

### ISSUE THREAD

**Issue by magnetophon** [magnetophon/DigiDrie#12]
different waveforms, retrigger, beat sync

**Comment by ryukau** [comment]
For waveform, is it possible to use naive wavetable in Faust? I could provide similar [UI used in CubicPadSynth](https://ryukau.github.io/VSTPlugins/manual/CubicPadSynth/img/cubicpadsynth.png) to directly draw waveform.

**Comment by ryukau** [comment]
I have a question around LFO and envelope.

Currently, `hbargraph` of LFO and envelope phases are provided for each macro. Are you planning to increase LFO and envelope that each macro has 4 LFO and 4 envelope? (This results in total 32 LFO and 32 envelope.)

I managed to build generated code for parameters, and next target is to generate UI code. I'd like to get better picture of how many DSP->UI elements are available.

**Comment by magnetophon** [comment]
Not sure if I understand the idea correctly:
I would create 64 faders per LFO, and just crossfade from nr1 to nr2 to nr3  ...   to nr64 and then loop back to nr1, right?

No problem, but I'm not yet convinced that this is the way to go. 

I was thinking of using 4 more CZ oscs as LFOs.
They would also get their phase from one master osc, but this master runs at much lower speeds, and can be synced to the beat.
By crossfading between octaves, you would get different subdivisions of the beat.

On the other hand, we might need to think about dropping features instead of adding them, to get the CPU usage under control.

What do you think?

**Comment by ryukau** [comment]
Yes. And crossfade part is done offline (or non-realtime). The synth I made just have refresh button to update LFO waveform.

CZ LFO is nice idea. It would adds more character, which I believe crucial for a synthesizer. I leave decision to you.

For dropping features, I agree with you. Though, I don't think we have to be hurry for it. I think benchmark helps to identify the candidates to drop. I'm currently focusing on UI, so I'll back this topic once I get the UI to working state.

**Comment by magnetophon** [comment]
> Are you planning to increase LFO and envelope that each macro has 4 LFO and 4 envelope? (This results in total 32 LFO and 32 envelope.)

I missed this question at first.

I think it's nice to have every LFO and envelope available for every macro.
I'm even thinking about removing "envelope -1" and using putting the overall volume under a modMixer.
What do you think?

As you know, we can always change the [number of LFO's](https://github.com/magnetophon/DigiDrie/blob/c5a3afdf415dec05631bf2e39c609b611ab79540/faust/lib/DigiDrie.lib#L943) and the [number of envelopes](https://github.com/magnetophon/DigiDrie/blob/c5a3afdf415dec05631bf2e39c609b611ab79540/faust/lib/DigiDrie.lib#L944).

If we have enough DSP available, that is.
And patience for the compile...  :)

**Comment by ryukau** [comment]
I'm not sure about moving volume control to macro without experimenting on it. A thing I'd like to avoid is unintentionally stop the sound. In this sense, I like simplicity around volumes in current implementation.

**Comment by magnetophon** [comment]
> A thing I'd like to avoid is unintentionally stop the sound.

Good point.
I'll leave it like it is.

**Comment by magnetophon** [comment]
I added CZ LFO's [here](https://github.com/magnetophon/DigiDrie/tree/LFO), but the synth doesn't compile anymore for me:

Even after hours, there is no output after:
```
end Constant propagation (duration : 0.0241549)
```

Does it work for you?

I have [reported this bug](https://github.com/grame-cncm/faust/issues/480) to the faust developers, but they, understandably, want a minimal code example with a bug.
The problem is that this kind of bug only appears in complex code...
